designs/0004-sparse-matrices.md

```diff
@@ -0,0 +1,387 @@
+  int<lower=1> nz_foo_row_index[K]; // Indices for nonzero elements
+  int<lower=1> nz_foo_col_index[K]; // Indices for nonzero elements
+
+  sparse_vector[N, non_zero_index] x;
+}
+parameters {
+  real<lower=0> rho;
+  real<lower=0> alpha;
+  real<lower=0> sigma;
+  // Defining sparse matrices in parameters needs the non-zero elements
+  sparse_matrix[N, M, nz_foo_row_index, nz_foo_col_index] foo;
+}
+transformed parameters {
+  // Non-zero elements are deduced by the operation on x
+  sparse_matrix[N, N] K = cov_exp_quad(x, alpha, rho);
+}
+
+```
+
+The size and non-zero indices for sparse matrices in the parameter block must be from either the data block or transformed data block. This is because Stan's I/O and posterior analysis infrastructure assumes the same full specification in each iteration of the model.
+
+## Full Example Model
+
+Below is a full example model that uses sparse vectors and matrices for data and parameters for a gaussian process.
+
+```stan
+data {
+  int<lower=1> N; // Vec size
+  int<lower=1, upper=N> K; // # of nonzero elements
+  int<lower=1> non_zero_index[K]; // Indices for nonzero elements
+  sparse_vector[N, non_zero_index] x;
+  vector[N] y;
+}
+transformed data {
+  sparse_vector[N] mu = rep_sparse_vector(0, N, non_zero_index); // [1]
+}
+parameters {
+  real<lower=0> rho;
+  real<lower=0> alpha;
+  real<lower=0> sigma;
+}
+model {
+  sparse_cholesky_matrix[N] L_K; // [2]
+  sparse_matrix[N, N] K = cov_exp_quad(x, alpha, rho);
+  real sq_sigma = square(sigma);
+
+  // Assign to pre-existing diagonal
+  for (n in 1:N)
+    K[n, n] = K[n, n] + sq_sigma;
+
+  L_K = cholesky_decompose(K);
+
+  rho ~ inv_gamma(5, 5);
+  alpha ~ std_normal();
+  sigma ~ std_normal();
+
+  y ~ multi_normal_cholesky(mu, L_K);
+}
+```
+
+- [1] Because Stan is told what values are true zeros vs. sparse zeros we can set this up fine.
+
+- [2] The `sparse_cholesky_matrix` is optional and holds the permutation of the sparse matrix after decomposition using the `EIGEN_SPARSEMATRIX_BASE_PLUGIN`. Permuting the rows of the sparse matrix before decomposition can often increase performance. Instead of having to recompute the permutation on each call to `cholesky_decompose`, after the first iteration we can keep the [`PermutationMatrix`](https://eigen.tuxfamily.org/dox/classEigen_1_1SimplicialCholeskyBase.html#aff1480e595a21726beaec9a586a94d5a) inside of the `sparse_cholesky_matrix` and reuse it on subsequent operations.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## I/O
+
+Data input formats should not need to change, less R and Python want a particular method that does not exist at this time.
+
+At the C++ level input data can be constructed to the sparse matrix through soemthing like what Ben did [here](https://github.com/stan-dev/rstanarm/blob/master/inst/include/csr_matrix_times_vector2.hpp#L18) for a revised `csr_matrix_times_vector`.
+
+```cpp
+int outerIndexPtr[cols+1];
```

Comment:
Why is this a pointer?

Reply:
The name is mostly because it's a pointer in the SparseMatrix class, name could change here though